### ISSUE TITLE
Substate Selection upon Subscription

### ISSUE THREAD

**Issue by Ben-G** [ReSwift/ReSwift#43]
As @Aleksion and I discussed, it would be great to be able to select a sub-state when subscribing to the store. This would allow us to programmatically ensure that the subscriber can only access a subset of the state.

Here's one idea of mine how we could potentially implement it, providing a closure that performs the sub select. We could typecheck that the closure returns the same type as is expected in `newState`.

    override func viewWillAppear(animated: Bool) {
        ...
        store.subscribe(self, { $0.navigationState })
    }

    func newState(state: NavigationState) {
        contacts = state.dataState.contacts
    }

@Aleksion also mentioned another approach using containers outside of View Controllers that would select and pass down a substate to a View Controller explicitly.

**Comment by raheelahmad** [comment]
I would vote for the second model, sort of like dependency injection.

I think there are two other related issues:

1. A subscriber is able to subscribe to any state
2. The state needs to be stored globally. This sounds more problematic to me. There is a tight coupling between state subscribers and this global state variable. Makes testing difficult etc.

Thinking out aloud: For every substate `SubState` we could have a corresponding `SubStateConsumer`

```swift
protocol SubStateConsumer {
    var state: SubState?
}
```

where the ViewController conforms to `SubStateConsumer` and its `state` property is set by the outside "container". The container itself holds the AppState private.

**Comment by Aleksion** [comment]
We were actually discussing an approach not unlike what you're proposing here.
In ReduxJs you have the concept of container components - which are the only components that know about the store and the whole state.

These would then pass down relevant functions and state to it's sub components.

This approach would make it easier to create isolated components - that could potentially be reused across projects. 

I kinda like the opportunities - we just have to figure out how to beat implement it and the potential downsides

**Comment by Ben-G** [comment]
This blog post contains some interesting insight into how Om (current and feature version) do and don't handle state subselects. Should provide some good insights:
http://blog.circleci.com/why-we-use-om-and-why-were-excited-for-om-next/

**Comment by Ben-G** [comment]
@Aleksion @raheelahmad My proposed solution would make substate selection look like this:

```
func selectSubstate(state: TestComplexAppState) -> (Int?, String?) {
        return (
            state.testValue,
            state.otherState?.name
        )
}
```

It allows you to select an arbitrary combinations of paths from your state. I'm actually pretty happy with this. Would be interested in your thoughts!

**Comment by Ben-G** [comment]
This is solved by #61.
